judo-runtime-core-dao-core/src/main/java/hu/blackbelt/judo/runtime/core/dao/core/processors/DeletePayloadDaoProcessor.java

```diff
@@ -224,12 +222,82 @@ void collectStaments(EClass entityType,
 
-
         // Make delete for all containment
         instanceGraph.getContainments().stream()
                 .forEach(i -> {
                     InstanceGraph<ID> containedGraph = i.getReferencedElement();
-                    collectStaments(i.getReference().getEReferenceType(), containedGraph, statements, instanceGraph, i.getReference());
+                    statmentCollector(i.getReference().getEReferenceType(), containedGraph, statements, instanceGraph, i.getReference());
+                });
+    }
+
+    private void checkMandatoryBackReferences(EClass entityType,
+                         InstanceGraph<ID> instanceGraph,
+                         Collection<Statement<ID>> statements,
+                         EReference container,
+                         Collection<InstanceValue<ID>> visited) {
+
+        InstanceValue<ID> instanceValue = InstanceValue.<ID>buildInstanceValue()
+                .type(entityType)
+                .identifier(instanceGraph.getId())
+                .build();
+
+        if (visited.stream().anyMatch(s -> Objects.equals(instanceValue, s))) {
+            log.debug("Circular delete found, stop collecting statements");
+            return;
+        }
+        visited.add(instanceValue);
+
+        // Check if there is any Remove Reference related to mandatory relation
+        Collection<InstanceReference<ID>> mandatoryReferencesToRemove = instanceGraph.getBackReferences().stream()
+                .filter(r -> r.getReference().getLowerBound() > 0 && !AsmUtils.annotatedAsTrue(r.getReference(), "reverseCascadeDelete")).collect(Collectors.toSet());
+        checkState(mandatoryReferencesToRemove.stream()
+                        .noneMatch(r -> statements.stream().noneMatch(s -> s instanceof DeleteStatement && Objects.equals(s.getInstance().getIdentifier(), r.getReferencedElement().getId()))),
+                "There are mandatory references that cannot be removed: " +
+                        " Type: " + entityType + " ID: " + instanceGraph.getId() + " Mandatory back references: " +
+                        mandatoryReferencesToRemove.stream().map(
+                                r -> r.getReference().getEContainingClass() + "#"
+                                        + r.getReference().getName() + " ID: "
+                                        + r.getReferencedElement().getId()).collect(toSet()));
+
+        Collection<EReference> processedReferences = Sets.newHashSet();
+        if (container != null) {
+            processedReferences.add(container);
+            if (container.getEOpposite() != null) {
+                processedReferences.add(container.getEOpposite());
+            }
+        }
+
+        instanceGraph.getReferences().stream()
+                .filter(r -> !processedReferences.contains(r.getReference()))
+                .filter(r -> r.getReference().getEOpposite() == null || !processedReferences.contains(r.getReference().getEOpposite()))
+                .forEach(r -> {
```

Comment:
Please replace stream foreach-s with the classic foreach. Especially because it contains a recursive call

Reply:
[0382708](https://github.com/BlackBeltTechnology/judo-runtime-core/pull/201/commits/0382708dba99d21fc499ed64fdba464108421a03)